File: problems/stack/stack.py
"""
This is a simple implementation of the Stack data structure
Use it as you please ^_^
"""

class Stack():
    """
    Simple LIFO Stack data structure
    """

    def __init__(self):
        """Constructor declaring the private variable"""
        self._items = []
    
    def is_empty(self):
        """Check the emptiness of the stack"""
        return len(self._items) == 0

    def size(self):
        """Get the number of items in the stack"""
        return len(self._items)

    def push(self, item):
        """Push an item to the stack"""
        self._items.append(item)

    def pop(self):
        """Pop an item from the end of the stack"""
        if self.is_empty():
            raise RuntimeError("Attempt to pop an empty Stack!")
        return self._items.pop()

    def show(self):
        """Show the contents of the stack"""
        res = "Stack (["
        if self.is_empty():
            print(res + "])")
            return
        for i in range(len(self._items)-1):
            res += str(self._items[i]) + ', '
        res += str(self._items[len(self._items)-1]) + "])"
        print(res)
        return


def main():
    stack = Stack()

    stack.show()          # Stack ([])
    stack.push(42)        # At this point, the stack looks like this: Stack ([42])
    stack.push(23)        # At this point, the stack looks like this: Stack ([42, 23])
    stack.push(5)         # At this point, the stack looks like this: Stack ([42, 23, 5])
    print(stack.size())   # 3
    stack.show()          # Stack ([42, 23, 5])
    print(stack.pop())    # 5
    stack.show()          # Stack ([42, 23])
    print(stack.pop())    # 23
    print(stack.size())   # 1
    stack.show()          # Stack ([42])
    print(stack.pop())    # 42
    stack.show()          # Stack ([])
    print(stack.pop())    # Underflow condition, should raise the RuntimeError exception and print "Attempt to pop an empty Stack!"


if __name__ == "__main__":
    main()

File: problems/stack/stack.go
package main

/*
An example of a stack being implemented in Go,
*/

import "fmt"

func main() {
	// Initalize the stack type
	stack := Stack{}

	stack.Show()              // Stack: []
	stack.Push(42)            // Adds to the stack, current stack = [42]
	stack.Push(23)            // Adds to the stack, current stack = [42 23]
	stack.Push(5)             // Adds to the stack, current stack = [42 23 5]
	fmt.Println(stack.Size()) // 3
	stack.Show()              // Stack: [42 23 5]
	fmt.Println(stack.Pop())  // 5
	stack.Show()              // Stack: [42 23]
	fmt.Println(stack.Pop())  // 23
	fmt.Println(stack.Size()) // 1
	stack.Show()              // Stack: [42]
	fmt.Println(stack.Pop())  // 42
	stack.Show()              // Stack: []
	fmt.Println(stack.Pop())  // panic: Attempt to pop an empty Stack!
}

// Struct to hold the stack items
type Stack struct {
	Items []int
}

// Show prints out the current stack
func (stack *Stack) Show() {
	fmt.Printf("Stack: %+v\n", stack.Items)
}

// Push adds an item to the current stack
func (stack *Stack) Push(item int) {
	stack.Items = append(stack.Items, item)
}

// Pop removes the last item from the stack
// fails if the stack is empty
func (stack *Stack) Pop() int {
	if !stack.IsEmpty() {
		poppedItem := stack.Items[len(stack.Items)-1]
		stack.Items = stack.Items[:len(stack.Items)-1]
		return poppedItem
	}
	panic("Attempt to pop an empty Stack!")
}

// Size returns the length of the current stack as int
func (stack *Stack) Size() int {
	return len(stack.Items)
}

// IsEmpty returns where the stack is empty or not as boolean
func (stack *Stack) IsEmpty() bool {
	return len(stack.Items) == 0
}
